stl/inc/xutility

```diff
@@ -4804,17 +4814,41 @@ template <class _FwdIt1, class _FwdIt2, class _Pr>
 bool _Is_permutation_unchecked(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2, _FwdIt2 _Last2, _Pr _Pred,
     forward_iterator_tag, forward_iterator_tag) {
     // test if [_First1, _Last1) == permuted [_First2, _Last2), using _Pred, arbitrary iterators
-    for (; _First1 != _Last1 && _First2 != _Last2; ++_First1, (void) ++_First2) {
+    for (;;) { // trim matching prefix
+        if (_First1 == _Last1) {
+            return _First2 == _Last2; // sequence equal
+        }
+
+        if (_First2 == _Last2) {

```

Comment:
Codegen for that appears worse: https://gcc.godbolt.org/z/oHGQ6r

Reply:
![image](https://user-images.githubusercontent.com/1544943/72564735-54931880-3865-11ea-91f9-098a39210da1.png)

Reply:
I've proven once again that I can't reason about MSVC's optimizer.

Reply:
FWIW LLVM says that those are identical: 
![image](https://user-images.githubusercontent.com/1544943/72567909-bf475280-386b-11ea-8784-ac75a3d7c495.png)

Reply:
Yes, so does GCC. I was expecting identical codegen for pointers, and better for less-analyzable iterators. 🤷‍♂

Reply:
That sounds like a worthy "please improve codegen for common usecase" ticket

Reply:
> That sounds like a worthy "please improve codegen for common usecase" ticket

I won't file bugs like that without a performance test showing the difference and I think it would be very difficult to meaningfully measure the difference here.